### PR TITLE
Store prometheus components in Instance

### PR DESCRIPTION
**Pull request by rfratto**
This doesn't do anything right now, but it sets up the groundwork for calling ApplyConfig against components from outside of the Run method. 

Related to #140.

### CODE REVIEW

File: pkg/prom/instance/instance.go [hunk 1]
```diff
@@ -306,47 +291,115 @@ func (i *Instance) Run(ctx context.Context) error {
...
 
+// initialize sets up the various Prometheus components with their initial
+// settings. initialize will be called each time the Instance is run. Prometheus
+// components cannot be reused after they are stopped so we need to recreate them
+// each run.
+//
+// Components are stored as struct members to allow other methods to access
+// components from different goroutines (as opposed to keeping them completely
+// defined within the Run method where nothing else can access them).
+func (i *Instance) initialize(ctx context.Context, reg prometheus.Registerer) error {
```
Comment (gotjosh) [inline pkg/prom/instance/instance.go:346]: You probably already know this but we need to stop the old components before we start new ones (e.g. `i.wal.Close()`).

 If you're planning to do that outside of this function it becomes a bit hard to follow the whole logic.

If you're planning to do it here, then the name of the function might be a bit deceiving (?)

What I'm trying to say is that depending on what you're thinking to do next there might be work to do here. I'll leave that your judgement.

File: pkg/prom/instance/instance.go [hunk 2]
```diff
@@ -306,47 +291,115 @@ func (i *Instance) Run(ctx context.Context) error {
...
+
+	// Setup the remote storage
+	remoteLogger := log.With(i.logger, "component", "remote")
+	i.remoteStore = remote.NewStorage(remoteLogger, reg, i.wal.StartTime, i.wal.Directory(), i.cfg.RemoteFlushDeadline, i.readyScrapeManager)
+	err = i.remoteStore.ApplyConfig(&config.Config{
+		GlobalConfig:       i.globalCfg,
+		RemoteWriteConfigs: i.cfg.RemoteWrite,
+	})
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed applying config to remote storage", "err", err)
```
Comment (gotjosh) [inline pkg/prom/instance/instance.go:372]: Do we need these log lines? From what I read the error will bubble up and be logged at the same log level.

File: pkg/prom/instance/instance.go [hunk 3]
```diff
@@ -306,47 +291,115 @@ func (i *Instance) Run(ctx context.Context) error {
...
+
+	i.storage = storage.NewFanout(i.logger, i.wal, i.remoteStore)
+
+	scrapeManager := newScrapeManager(log.With(i.logger, "component", "scrape manager"), i.storage)
+	err = scrapeManager.ApplyConfig(&config.Config{
+		GlobalConfig:  i.globalCfg,
+		ScrapeConfigs: i.cfg.ScrapeConfigs,
+	})
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed applying config to scrape manager", "err", err)
```
Comment (gotjosh) [inline pkg/prom/instance/instance.go:384]: Ditto.

File: pkg/prom/instance/instance.go (hunk 1 above)
Comment (rfratto) [reply pkg/prom/instance/instance.go:346]: Right, the components are stopped at the end of Run. The initialize method was pulled out to make it easier to grab the mutex. Would more comments or better naming be ideal here?

Comment (gotjosh) [reply pkg/prom/instance/instance.go:346]: Ah! I put it together now, this is only for creating the new components but not starting/stopping them as that is still done within `Run` via the run group.

Why do you need a lock then? And if you do, shouldn't you be holding it for the whole duration until right before you call `rg.Run()`?

Comment (rfratto) [reply pkg/prom/instance/instance.go:346]: We need a lock because another goroutine might [request the active targets](https://github.com/rfratto/agent/blob/032826b89b1aea62a2f10f346c630f98fbd3b37f/pkg/prom/instance/instance.go#L395) at any time (i.e., the uses calls the new `/api/agent/v1/targets` API).

It doesn't matter if the components aren't run yet, we just need to prevent a concurrent read/write to the pointers.

Comment (gotjosh) [reply pkg/prom/instance/instance.go:346]: 🤦 sorry, it all makes sense now - I read this all wrong.
